subsys/settings/settings_fcb.c

```diff
@@ -0,0 +1,619 @@
+
+#include <settings/settings.h>
+#include "settings_priv.h"
+
+struct settings_fcb_read_fn_arg {
+	struct fcb_entry_ctx *entry_ctx;
+	off_t off;
+	u8_t rbs;
+};
+
+static int settings_fcb_load(struct settings_store *cs);
+static int settings_fcb_save(struct settings_store *cs, const char *name,
+			     const char *value, size_t val_len);
+
+static struct settings_store_itf settings_fcb_itf = {
+	.csi_load = settings_fcb_load,
+	.csi_save = settings_fcb_save,
+};
+
+int settings_fcb_dst(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_dst_register(&cf->cf_store);
+
+	return 0;
+}
+
+int settings_fcb_src(struct settings_fcb *cf)
+{
+	cf->cf_store.cs_itf = &settings_fcb_itf;
+	settings_src_register(&cf->cf_store);
+
+	return 0;
+}
+
+static int settings_fcb_read(struct fcb_entry_ctx *entry_ctx, off_t off,
+			     char *buf, size_t *len)
+{
+
+	if (off >= entry_ctx->loc.fe_data_len) {
+		return -EINVAL;
+	}
+
+	if ((off + *len) > entry_ctx->loc.fe_data_len) {
+		*len = entry_ctx->loc.fe_data_len - off;
+	}
+
+	return flash_area_read(entry_ctx->fap,
+			       FCB_ENTRY_FA_DATA_OFF(entry_ctx->loc) + off,
+			       buf, *len);
+}
+
+static ssize_t settings_fcb_read_fn(void *back_end, void *data, size_t len)
+{
+	int rc, rem;
+	struct settings_fcb_read_fn_arg *rd_fn_arg;
+	char temp_buf[16]; /* buffer for fit read-block-size requirements */
+	size_t len_read, tmp_len;
+	off_t off;
+	u8_t *data8 = (u8_t *)data;
+
+	len_read = 0;
+	rd_fn_arg = (struct settings_fcb_read_fn_arg *)back_end;
+	off = rd_fn_arg->off;
+
+	/* Take care of alignment */
+	rem = rd_fn_arg->off % rd_fn_arg->rbs;
+
+	if (rem) {
+		/* Read from aligned offset, keep only required data */
+		off -= rem;
+		len_read = min(rd_fn_arg->rbs - rem, len);
+		tmp_len = rd_fn_arg->rbs;
+		rc = settings_fcb_read(rd_fn_arg->entry_ctx, off, temp_buf,
+				       &tmp_len);
+		if (rc) {
+			return rc;
+		}
+		memcpy(data8, temp_buf+rem, len_read);
+		len -= len_read;
+		data8 += len_read;
+		off += rd_fn_arg->rbs;
+	}
+
+	/* If there is more data read it */
+	if (len) {
+		tmp_len = len;
+		rc = settings_fcb_read(rd_fn_arg->entry_ctx, off, data8,
+				       &tmp_len);
+		if (rc) {
+			return rc;
+		}
+
+		len_read += tmp_len;
+	}
+
+	return len_read;
+}
+
+static int settings_fcb_load(struct settings_store *cs)
+{
+	struct settings_fcb *cf = (struct settings_fcb *)cs;
+	struct fcb_entry_ctx loc;
+	struct settings_fcb_read_fn_arg read_fn_arg;
+	struct settings_handler *ch;
+	char name[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN];
+	char *name_argv[SETTINGS_MAX_DIR_DEPTH];
+	int rc, name_argc;
+	size_t len_name, len_data;
+	u8_t rbs;
+
+	rbs = flash_area_align(cf->cf_fcb.fap);
+	loc.fap = cf->cf_fcb.fap;
+	loc.loc.fe_sector = NULL;
+	loc.loc.fe_elem_off = 0;
+
+	while (fcb_getnext(&cf->cf_fcb, &loc.loc) == 0) {
+
+		len_name = min(SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1,
+				loc.loc.fe_data_len);
+
+		rc = settings_fcb_read(&loc, 0, name, &len_name);
```

Comment:
name begin address is always aligned while end might not.
I recommend to Introduce intermediate function for read without care about alignment and use it elsewhere.